File: Models/OrderStats.cs
using System;
using System.Collections.Generic;

namespace KajfestPOS.Models
{
    public class OrderStats
    {
        internal DateTimeOffset EarliestOrderCreated { get; set; }
        public DateTimeOffset PeriodStart => EarliestOrderCreated.AddTicks(-(EarliestOrderCreated.Ticks % TimeSpan.TicksPerHour));
        public int Orders { get; set; }
        public decimal Total { get; set; }
        public List<PaymentStats> Payments { get; set; }
        public List<ProductStats> Products { get; set; }
    }

    public class PaymentStats
    {
        internal DateTimeOffset EarliestOrderCreated { get; set; }
        internal DateTimeOffset PeriodStart => EarliestOrderCreated.AddTicks(-(EarliestOrderCreated.Ticks % TimeSpan.TicksPerHour));
        public PaymentMethod Method { get; set; }
        public int Payments { get; set; }
        public decimal Total { get; set; }
    }

    public class ProductStats
    {
        internal DateTimeOffset EarliestOrderCreated { get; set; }
        internal DateTimeOffset PeriodStart => EarliestOrderCreated.AddTicks(-(EarliestOrderCreated.Ticks % TimeSpan.TicksPerHour));
        public int ProductId { get; set; }
        public string ProductName { get; set; }
        public int Orders { get; set; }
        public int Quantity { get; set; }
        public decimal Total { get; set; }
    }
}

File: src/api/order-stats.ts
import { DateTime } from "luxon";
import { jsonProperty } from "ur-json";
import { dateTimeConverter } from "./converters/date-time-converter";
import { bigConverter } from "./converters/big-converter";
import { Big } from "big.js";
import { PaymentMethod } from "./payment";

class PaymentStats {
    @jsonProperty()
    method!: PaymentMethod;

    @jsonProperty()
    payments!: number;

    @jsonProperty({ converter: bigConverter })
    total!: Big;
}

class ProductStats {
    @jsonProperty()
    productId!: number;

    @jsonProperty()
    productName!: string;

    @jsonProperty()
    orders!: number;

    @jsonProperty()
    quantity!: number;

    @jsonProperty({ converter: bigConverter })
    total!: Big;
}

export class OrderStats {
    @jsonProperty({ converter: dateTimeConverter })
    periodStart!: DateTime;

    @jsonProperty()
    orders!: number;

    @jsonProperty({ converter: bigConverter })
    total!: Big;

    @jsonProperty({type: PaymentStats})
    payments!: PaymentStats[];

    @jsonProperty({type: ProductStats})
    products!: ProductStats[];
}